### ISSUE TITLE
Group cosets: implement custom `in`, improve iterator

### ISSUE THREAD

**Issue by fingolfin** [oscar-system/Oscar.jl#4289]
Right now `in` for group cosets resorts to iterating over the coset, which is very inefficient.

We should provide a custom method for this. I would not dispatch to GAP's, but instead implement this "directly": i.e. to test `g in Hx`, test if `g/x in H` and similar for `g in xH` (so that we avoid creating `GapObj` for those).

---

Speaking of iterators, the `iterate` implementation for `GroupCoset` currently directly wraps a GAP coset iterator. But (a) this requires creating a `GapObj` (with extra overhead for left cosets). And furthermore on the GAP side, the iterator is a wrapper for an enumerator. 

Instead, the Julia iterator for a coset `C` could be based on the *group* iterator for `C.H`.

---

Lastly, contrast
```
Base.IteratorSize(::Type{<:GAPGroup}) = Base.SizeUnknown()
Base.IteratorSize(::Type{PermGroup}) = Base.HasLength()
```
versus
```
Base.IteratorSize(::Type{<:GroupCoset}) = Base.SizeUnknown()
```

I suggest we use something like this:
```
Base.IteratorSize(::Type{<:GroupCoset{T}}) where T = Base.IteratorSize(T)
```

And then I assume similar for `GroupDoubleCoset`.

**Comment by ThomasBreuer** [comment]
For a coset `Hg` or `gH` of type `GroupCoset{T, S}`, `T` is not necessarily the type of `H`.
If we want to set `Base.IteratorSize` for cosets depending on `Base.IteratorSize` for the type of `H` then I think we have to change the parameterization to `GroupCoset{TG, TH, S}` where `TG` and `TH` are the types of `G` and `H`, respectively.

Concerning the iterator for double cosets, one solution would be to use the functionality of GAP's `RepresentativesContainedRightCosets` (which is also used by GAP's membership test for double cosets), and to delegate to the iterator for right cosets.

**Comment by ThomasBreuer** [comment]
Where are we now:

With #4302, we get
- `in` for cosets delegating to `in` for groups, and `in` for double cosets delegating to `GAP`
- an extended parameterization of the `GroupCoset` type such that the idea to get a better `Base.IteratorSize` for cosets works

Concerning double cosets, we would need more GAP functionality on the Oscar side in order to replace the delegations to GAP by something better. For both iteration and membership tests for double cosets, my idea would be to split the double coset into right cosets and to delegate to these object, as is done in GAP. For that, GAP's `RepresentativesContainedRightCosets` could be exported to Oscar. (I am not sure whether GAP's approach using canonical coset representatives is generically a good idea, it defaults to unpacking right cosets and taking the minimal element.)

**Comment by fingolfin** [comment]
I am actually fine with leaving double cosets as they are right now. My mention of double cosets above actually was only referring to adding/modifying the `Base.IteratorSize` for double cosets.

So for me this issue was resolved by PR #4302.
